ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/target/TargetDefinition.java

```diff
@@ -899,10 +900,21 @@ public boolean isContentEquivalent(ITargetDefinition definition) {
 				isArgsNullOrEqual(getVMArguments(), definition.getVMArguments()) && //
 				Objects.equals(getJREContainer(), definition.getJREContainer()) && //
 				Arrays.equals(getIncluded(), definition.getIncluded()) && //
-				Arrays.equals(getTargetLocations(), definition.getTargetLocations()) && //
+				unorderedArrayEqualsSerialized(getTargetLocations(), definition.getTargetLocations()) && //
 				Arrays.equals(getImplicitDependencies(), definition.getImplicitDependencies());
 	}
 
+	private boolean unorderedArrayEqualsSerialized(ITargetLocation[] one, ITargetLocation[] two) {
+		if (one.length != two.length) {
+			return false;
+		}
+		Set<String> oneXml = new HashSet<>(
+				Stream.of(one).filter(l -> l != null).map(ITargetLocation::serialize).toList());
+		Set<String> twoXml = new HashSet<>(
+				Stream.of(two).filter(l -> l != null).map(ITargetLocation::serialize).toList());
```

Comment:
What do you think about clean up the xml e.g. replace all \r \n\t by empty string?

Or even parse it with a dom parser and then process it on the dom level, e.g. order of attributes should not matter also.

Reply:
Uhm, that feels like overengineering:
AFAICS, a difference in the serialized string without a difference in the real content only happens if someone really opened the editor, performing some changes _and reverting_ them. In this case I don't think writing the file again or reloading target platform state really is a problem.

And the newline thingy is a different problem anyway: https://github.com/eclipse-pde/eclipse.pde/issues/1059

Reply:
> What do you think about clean up the xml e.g. replace all \r \n\t by empty string?

Thinking more about this, also having https://github.com/eclipse-pde/eclipse.pde/issues/1059 in mind, brings me to the idea that serialize() probably isn't the best approach to write into XML files or even create XML strings, because it doesn't allow to use some specific (XML) Writer doing whatever formatting is wanted (regarding line breaks, indentation, ...).
Heck, IMHO the data object shouldn't be aware at all of whether the result should be XML/JSON/Binary/Whatever, nor File/String/Stream/Whatever, and any particular formatting, no?

Reply:
Yes but as it is API we can't easily change that... so we can only do our best here I also planned to rework the API but its not top priority at them moment as it works so far.